Add data plane tests for nixl_ep

eranrs opened the pull request:
Test categories:
- Functional tests: connection management, dispatch, combine, masking, end-to-end
- Performance tests: control plane latency, data plane throughput
- Bug reproductions: documented known issues (BUG-01 through BUG-05)
- Unit tests: individual component validation

Infrastructure:
- Multi-process test runner with GPU-per-rank assignment
- TCP-based rank server for process coordination
- Results collector for CI/CD integration
- Pytest marks for selective test execution

Documentation:
- README files for each test category
- Known issues and workarounds documented
- Multi-node testing considerations in .cursorrules

## What?
This PR adds a comprehensive pytest-based test suite for the nixl_ep example (Expert-Parallel communication buffer). The test suite includes:
Functional tests: Connection management, dispatch, combine, masking, and end-to-end workflows
Performance tests: Control plane latency and data plane throughput benchmarks
Bug reproductions: Documented known issues (BUG-01 through BUG-05) for tracking and regression testing
Unit tests: Individual component validation
Supporting infrastructure:
Multi-process test runner with GPU-per-rank assignment
TCP-based rank server for process coordination
Results collector for CI/CD integration
Pytest marks for selective test execution
## Why?
The nixl_ep example currently lacks automated tests, making it difficult to:
Verify correctness after code changes
Track performance regressions
Document and reproduce known issues
Validate multi-GPU configurations
This test suite enables CI/CD integration and provides a foundation for quality assurance as the codebase evolves.
## How?
Tests are located in test/python/nixl_ep_tests/ with the following structure:
functional/ - Multi-process tests using mp.spawn with 8 GPUs
perf/ - Benchmarks for control/data plane operations
bugs/ - Reproduction cases for known issues
unit/ - Single-process component tests
utils/ - Shared test infrastructure (mp_runner, rank_server, etc.)
Run with: python -m pytest test/python/nixl_ep_tests/ -v

itayalroy committed:
nixl_ep: Skip invalidate MD with TCPStore

Signed-off-by: Itay Alroy <ialroy@nvidia.com>

itayalroy committed:
nixl_ep: Increase wait time for tcp store md exchange

Signed-off-by: Itay Alroy <ialroy@nvidia.com>

itayalroy committed:
nixl_ep: Migrate elastic.py to TCPStore

This commit migrates elastic.py to TCPStore-based metadata exchange
instead of ETCD.

Note that this TCPStore instance can be re-used for rank management
(instead of the custom TCP based implementation we currently have)
but distributed rank management performs worse in scale (tested
with 64 concurrent get_rank() calls) so rank server will stay
for now.

Signed-off-by: Itay Alroy <ialroy@nvidia.com>

itayalroy committed:
Use tcp_store's multi_get to fetch all MDs at once

Signed-off-by: Itay Alroy <ialroy@nvidia.com>

eranrs committed:
test(nixl_ep): add data plane performance test

Minimal performance test for NIXL EP Buffer data plane operations:
- dispatch throughput/latency (dispatch only)
- combine throughput/latency (one dispatch, many combines)
- e2e throughput/latency (dispatch + combine cycles)

Usage: python3 test_data_plane.py --num-processes=8 --mode=e2e

Includes:
- mp_runner.py: Multi-process test runner with GPU/UCX coordination
- rank_server.py: TCP-based coordination server for distributed tests

eranrs committed:
test(nixl_ep): add control plane performance test

Adds test_control_plane.py to measure latency of:
- Buffer initialization (init)
- connect_ranks()
- disconnect_ranks()
- destroy()
- Full cycle (init → connect → disconnect → reconnect → destroy)

Supports multiple expert counts per run and both IPC and RDMA backends.

eranrs committed:
feat(test): add --use-tcp-store flag to control plane test

Adds support for using TCPStore instead of etcd for metadata exchange
during control plane performance tests. Uses port offset (+1000) to
avoid conflicts with torchrun's default port.

Usage: python3 test_control_plane.py --use-tcp-store --nvlink-backend=none

eranrs committed:
fix(test): skip etcd check when using TCPStore

When --use-tcp-store is passed, skip the etcd availability check
since metadata exchange is done via TCPStore instead.

eranrs committed:
fix(test): pass use_tcp_store to test function via kwargs

The use_tcp_store parameter was captured as a named argument but not
passed through to the child processes, causing them to default to
use_tcp_store=False and hang waiting for etcd metadata.

eranrs committed:
refactor(test): adopt PR 1155 TCPStore pattern

- Add store_group.py helper from PR 1155
- Start dedicated TCPStore server process in mp_runner
- Use create_client_store() in workers (no world_size needed)
- Cleaner separation: master server + client workers
- Follows ai-dynamo/nixl PR 1155 elastic test pattern

eranrs committed:
Fix TCPStore metadata exchange in mp_runner

- Add use_tcp_store parameter to setup_worker_environment()
- Only set NIXL_ETCD_ENDPOINTS when NOT using TCPStore
- This prevents C++ code from activating etcd path when TCPStore is requested
- Matches elastic.py implementation from pr-1155 branch
- Fixes control plane tests to work with --use-tcp-store flag

eranrs committed:
feat(test): add multi-node support using WORLD_SIZE and RANK env vars

- Use standard PyTorch distributed env vars (WORLD_SIZE, RANK, MASTER_ADDR)
- CLI flags (--world-size, --rank, --master-addr) override env vars
- WORLD_SIZE = number of nodes (not total ranks)
- num-processes = GPUs per node
- RANK 0 = master node (runs TCPStore and rank server)
- RANK > 0 = worker nodes (connect to master)
- Validation: error if MASTER_ADDR not set for workers
- Calculate global rank = node_rank * procs_per_node + local_rank
- Backward compatible: defaults to single-node (WORLD_SIZE=1, RANK=0)

Example multi-node usage:
  # Master (node 0)
  WORLD_SIZE=2 RANK=0 MASTER_ADDR=node0 python3 test_control_plane.py --num-processes=8 --use-tcp-store

  # Worker (node 1)
  WORLD_SIZE=2 RANK=1 MASTER_ADDR=node0 python3 test_control_plane.py --num-processes=8 --use-tcp-store

eranrs committed:
fix(test): ensure only master node handles etcd/TCPStore/rank-server

- Only master node (RANK=0) checks/cleans etcd state
- Only master node starts TCPStore server
- Only master node starts rank coordination server
- Worker nodes connect to master's services
- Add clear logging for multi-node master/worker role
- Prevents conflicts when multiple nodes try to manage shared services

eranrs committed:
docs(test): add multi-node usage documentation

- Document WORLD_SIZE, RANK, MASTER_ADDR environment variables
- Provide examples for master and worker node setup
- Explain CLI flag alternatives (--world-size, --rank, --master-addr)
- Clarify that WORLD_SIZE = number of nodes, not total ranks
- Add note that master node runs TCPStore/rank server
- Recommend --use-tcp-store to avoid etcd dependency

eranrs committed:
fix(test): use correct RankClient API for rank registration

- RankClient uses get_rank() not register_rank()
- Multi-node: calculate deterministic global_rank = RANK * num_processes + local_rank
- Single-node: use server-assigned global_rank (auto-incrementing)
- Prevents non-deterministic rank assignment in multi-node setup
- Fixes AttributeError: 'RankClient' object has no attribute 'register_rank'

eranrs committed:
fix(test): inherit UCX_TLS env var for RDMA multi-node support

- Spawned worker processes weren't inheriting UCX_TLS from parent shell
- Now explicitly set UCX_TLS=rc_mlx5,dc_mlx5,tcp if not already set
- Critical for cross-node RDMA communication in multi-node tests

eranrs committed:
fix(test): pre-set UCX_TLS with RDMA transports and cuda_ipc exclusion

- Set UCX_TLS=rc_mlx5,dc_mlx5,tcp,^cuda_ipc before Buffer creation
- Prevents buffer.py from overwriting with only ^cuda_ipc
- Ensures RDMA transports are enabled for cross-node communication

eranrs committed:
debug: add UCX_TLS logging to diagnose metadata issue

eranrs committed:
fix(test): always use rank server for rank assignment

- Simplified rank assignment: always use rank_server.get_rank()
- Works identically for single-node and multi-node
- Follows elastic.py pattern: ranks assigned sequentially (0,1,2,...)
- Each Buffer gets unique global_rank for agent name
- Ensures unique metadata in etcd/TCPStore
- torch_rank used for CUDA device (0-7 per node)

eranrs committed:
fix(test): use master_addr for etcd in multi-node setup

- etcd must be accessible from all nodes via master IP
- Auto-detect when etcd_server is localhost but master_addr is not
- Override to use http://master_addr:2379 for multi-node
- Matches elastic.py pattern: --etcd-server=MASTER_IP:2379
- Critical for cross-node metadata exchange

eranrs committed:
Add debug logging to track UCX_TLS and Buffer parameters

eranrs committed:
Add debug logging to trace NIXL_ETCD_ENDPOINTS setting

eranrs committed:
CRITICAL FIX: Restore UCX_TLS after Buffer.__init__() for multi-node

Buffer.__init__() sets UCX_TLS='^cuda_ipc' when nvlink_backend!='nixl'.
This overwrites RDMA transports needed for multi-node communication.
UCX initialization happens in update_memory_buffers() -> init() -> _nixl_agent_init().
We must restore RDMA transports AFTER Buffer() but BEFORE update_memory_buffers().

eranrs committed:
Revert UCX_TLS explicit setting - let buffer.py handle it

UCX_TLS=^cuda_ipc means 'use all transports except cuda_ipc'.
UCX will auto-detect RDMA transports (rc_mlx5, dc_mlx5, tcp).
Don't override what buffer.py sets.

eranrs committed:
Add debug logging following elastic.py pattern

Log UCX_NET_DEVICES and NIXL_ETCD_ENDPOINTS right before Buffer creation
to verify environment is set correctly for multi-node.

eranrs committed:
Align control plane tests with elastic.py pattern (PR 1155)

- Pass tcp_store_group to Buffer() constructor (like elastic.py)
- Remove set_tcp_store_group() calls (no longer needed)
- Cherry-picked 4 commits from PR 1155:
  * Skip invalidate MD with TCPStore
  * Increase wait time for tcp store md exchange
  * Migrate elastic.py to TCPStore
  * Use tcp_store's multi_get to fetch all MDs at once

eranrs committed:
Add TCP fallback NICs to UCX_NET_DEVICES for cross-node

Copy elastic.py pattern: append IPoIB (InfiniBand) TCP interfaces
to UCX_NET_DEVICES for cross-node communication fallback.

Without TCP NICs, UCX can't establish active messages transport
between nodes, causing 'no active messages transport' errors.

eranrs committed:
Fix multi-node result collection and add topology debug logging

- Each node now checks for its own expected global rank range
  (Master: 0-7, Worker: 8-15, etc.)
- Previous bug: All nodes checked for ranks 0-num_processes, causing
  worker to think master ranks died
- Add debug logging to show discovered GPU-NIC topology per node

eranrs committed:
Simplify rank naming to match elastic.py pattern

Make rank assignment consistent with elastic.py:
- local_rank: GPU index on this node (0-7) from rank server
  → Used for GPU/NIC selection (CUDA_VISIBLE_DEVICES, UCX_NET_DEVICES)
- global_rank: Unique rank across all nodes (0-15 for 2 nodes) from rank server
  → Used for Buffer(rank=...)
- torch_rank: Spawn index (internal to torch.multiprocessing, not used)

Previous bug: Used torch_rank for GPU/NIC selection instead of local_rank,
causing GPU 0 to get GPU 4's NIC, etc.

This matches the simple, clear pattern in elastic.py.

eranrs committed:
Add local_rank to debug logging

Show both global rank and local_rank in debug output to verify
rank server assignment is working correctly.

eranrs committed:
Fix NameError: pass local_rank to _run_single_op

The debug logging uses local_rank but it wasn't passed as a parameter
to _run_single_op function. Added local_rank parameter and argument.

eranrs committed:
Show per-node results instead of master-only

- Modified log_cycle_results() to display results on each node
- Workers now show their own rank results (e.g., ranks 8-15)
- Master shows its rank results (e.g., ranks 0-7)
- Added rank range display to clarify which ranks each node collected

eranrs committed:
Add 60s retry/wait for workers to connect to master services

- Workers now wait up to 60 seconds for TCPStore and rank server
- Added _wait_for_tcp_service() helper with retry logic
- Allows starting all workers first, then master (more flexible startup order)
- Replaces fixed 2-second sleep with proper availability checking

eranrs committed:
Revert "Add 60s retry/wait for workers to connect to master services"

This reverts commit 348fa6cd6adbe1a43a63ff12e23fd05ba07031d3.

eranrs committed:
Add polling-based wait for worker nodes to connect to master servers

- Add wait_for_server() method to RankClient with exponential backoff
- Add wait_for_tcp_port() helper function for TCPStore
- Replace fragile time.sleep(2.0) with proper server readiness checks
- Workers now poll until rank server/TCPStore are ready (60s timeout)
- Includes progress logging for debugging multi-node startup

eranrs committed:
Fix: Workers should not clear barriers - only master should

Workers calling clear_barriers() after connecting was clearing barriers
that master's processes were already waiting on, causing barrier timeouts.

eranrs committed:
Fix: Use deterministic rank assignment based on node_rank and local process index

Previous approach used connection-order rank assignment from rank server,
which caused unpredictable rank distribution across nodes. Now:
- global_rank = node_rank * num_processes + local_rank
- Node 0 gets ranks 0-7, Node 1 gets ranks 8-15, etc.
- Removes dependency on rank server for rank assignment (still used for barriers)

eranrs committed:
Add explicit logging when workers confirm master is alive

Workers now log:
- ✓ TCPStore ready at <master>:<port>
- ✓ Master is alive! Connected to rank server at <master>:<port>

eranrs committed:
Add [Node X] prefix to all log messages for easier multi-node debugging

All log messages from run_multiprocess_test now start with [Node X]
where X is the node's rank (0=master, 1-N=workers)

eranrs committed:
Add [Node X] prefix to wait_for_tcp_port and wait_for_server log messages

All server readiness polling messages now include the node prefix for
consistent multi-node debugging output

eranrs committed:
Add [Node X] prefix to all test results and headers

- Pass node_rank from mp_runner to test function via kwargs
- Configure logger formatter in test function with [Node X] prefix
- All output including results tables now prefixed for multi-node debugging

eranrs committed:
Add [Node X] prefix to main() header and results output

Set logger formatter early in main() so all output including
the test configuration header and results tables have the prefix

eranrs committed:
Fix double [Node X] prefix in log messages

- Use logger formatter instead of manual prefix in mp_runner.py
- Remove log_prefix parameter from wait_for_tcp_port and wait_for_server
- All logs now get prefix from formatter automatically

eranrs committed:
Add multi-node support to test_data_plane.py

- Add --world-size, --rank, --master-addr parameters
- Add --use-tcp-store for TCPStore metadata exchange
- Add [Node X] prefix to all log messages
- Support TCPStore in _run_data_plane_test function
- Consistent with test_control_plane.py interface

eranrs committed:
Fix missing [Node X] prefix on 'Running:' log message

The \n at the start of the log message caused the prefix to appear
after the newline. Using separate logger.info('') for blank line.

eranrs committed:
Fix store_group import in test_data_plane.py

Was using 'from store_group import store_group' which failed.
Changed to 'import store_group' to match test_control_plane.py

eranrs committed:
Remove test_control_plane.py - focus PR on data plane tests only

eranrs committed:
Fix pre-commit issues: isort, flake8 (unused imports/variables)

itayalroy commented on test/python/nixl_ep_perf/mp_runner.py:
```diff
@@ -0,0 +1,609 @@
+_RANK_SERVER_PORT: int = 9998
+
+
+def discover_gpu_nic_topology() -> Optional[Dict[int, str]]:
```
Is this still needed? UCX issues should be fixed already

itayalroy commented on test/python/nixl_ep_perf/mp_runner.py:
```diff
@@ -0,0 +1,609 @@
+        env = os.environ.copy()
+        env["ETCDCTL_API"] = "3"
+
+        for prefix in ["/nixl", "nixl"]:
```
This is dependent on NIXL's internal impl, I don't think you should rely on that, why not clean all keys?

eranrs replied:
Will clean all

itayalroy commented on test/python/nixl_ep_perf/README.md:
```diff
@@ -0,0 +1,160 @@
+- RANK>0 are workers (connect to master)
+- Use `--use-tcp-store` to avoid etcd dependency
+
+### Example Output
```
Can you share data plane example output here as well?

itayalroy reviewed (commented):
How does dispatch/combine test results compare to what we see with elastic.py? can you share some results?

eranrs commented on test/python/nixl_ep_perf/test_data_plane.py:
```diff
@@ -0,0 +1,443 @@
+    if other_ranks:
+        torch.cuda.synchronize()
+        buffer.connect_ranks(other_ranks)
+        torch.cuda.synchronize()
```
I am removing them both to be similar to elastic.py, will verify in tests

eranrs committed:
test(nixl_ep): address PR review comments

- Remove double torch.cuda.synchronize() around connect_ranks
- Clean all etcd keys instead of just nixl/* prefix
- Simplify etcd_server logic to always use master_addr
- Remove manual wait_for_tcp_port, use TCPStore built-in timeout
- Invert NIC discovery: default=skip, add --discover-nics flag
- Invert metadata exchange: default=TCPStore, add --use-etcd flag

eranrs committed:
fix(nixl_ep): ensure TCPStore server is ready before spawning workers

- Add wait_for_tcp_port for both master and worker nodes
- Prevents race condition where workers try to connect before server is ready
- Fixes timeout errors in single-node TCPStore tests

eranrs committed:
fix(nixl_ep): keep TCPStore object alive to prevent garbage collection

- Store returned object in variable to prevent Python GC
- Without this, TCPStore server would immediately shut down
- Fixes 'TCP port not ready' timeout error

eranrs committed:
style(nixl_ep): fix flake8 unused variable warning

- Add noqa comment for intentionally unused _store variable
- Variable must be kept alive to prevent garbage collection

eranrs committed:
chore: revert changes to examples/ and pyproject.toml

Focus PR scope on test/python/nixl_ep_perf only

eranrs committed:
style(nixl_ep): fix flake8 E226 and import order

- Add spacing around arithmetic operator (world_size - 1)
- Fix import order to be alphabetical (nixl_ep after numpy/torch)
- Update copyright year to 2025-2026

eranrs committed:
chore(nixl_ep): update copyright to 2025-2026 and fix import order

- Update copyright headers for files modified in 2026
- Fix import order: nixl_ep before numpy/torch (CI requirement)

itayalroy reviewed (approved):
Given the size of this framework, my review focused on the high-level design. Overall it looks good.

eranrs committed:
Merge branch 'main' into adding_tests
